pkg/api/meta/builds_test.go

```diff
@@ -0,0 +1,249 @@
+					{From: imageRefValue("test-1")},
+					{From: imageRefValue("test-2")},
+					{From: imageRefValue("test-3")},
+				}},
+			}},
+			args: args{fn: func(ref *kapi.ObjectReference) error {
+				if ref.Name == "test-2" {
+					ref.Name = "test-4"
+				}
+				return nil
+			}},
+			wantSpec: &buildapi.CommonSpec{
+				Source: buildapi.BuildSource{Images: []buildapi.ImageSource{
+					{From: imageRefValue("test-1")},
+					{From: imageRefValue("test-4")},
+					{From: imageRefValue("test-3")},
+				}},
+			},
+		},
+		{
+			name: "mutates only changed references",
+			fields: fields{
+				spec: &buildapi.CommonSpec{
+					Source: buildapi.BuildSource{Images: []buildapi.ImageSource{
+						{From: imageRefValue("test-1")},
+						{From: imageRefValue("test-2")},
+						{From: imageRefValue("test-3")},
+					}},
+				},
+				oldSpec: &buildapi.CommonSpec{
+					Source: buildapi.BuildSource{Images: []buildapi.ImageSource{
+						{From: imageRefValue("test-1")},
+						{From: imageRefValue("test-3")},
+					}},
+				},
+			},
+			args: args{fn: func(ref *kapi.ObjectReference) error {
+				if ref.Name != "test-2" {
+					t.Errorf("did not expect to be called for existing reference")
+				}
+				ref.Name = "test-4"
+				return nil
+			}},
+			wantSpec: &buildapi.CommonSpec{
+				Source: buildapi.BuildSource{Images: []buildapi.ImageSource{
+					{From: imageRefValue("test-1")},
+					{From: imageRefValue("test-4")},
+					{From: imageRefValue("test-3")},
+				}},
+			},
+		},
+		{
+			name: "skips when docker reference unchanged",
+			fields: fields{
+				spec: &buildapi.CommonSpec{
+					Strategy: buildapi.BuildStrategy{
+						DockerStrategy: &buildapi.DockerBuildStrategy{From: imageRef("test")},
+					},
+				},
+				oldSpec: &buildapi.CommonSpec{
+					Strategy: buildapi.BuildStrategy{
+						DockerStrategy: &buildapi.DockerBuildStrategy{From: imageRef("test")},
+					},
+				},
+			},
+			args: args{fn: func(ref *kapi.ObjectReference) error {
+				t.Errorf("should not have called mutator")
+				return nil
+			}},
+			wantSpec: &buildapi.CommonSpec{
+				Strategy: buildapi.BuildStrategy{
+					DockerStrategy: &buildapi.DockerBuildStrategy{From: imageRef("test")},
+				},
+			},
+		},
+		{
+			name: "skips when docker reference unchanged",
+			fields: fields{
+				spec: &buildapi.CommonSpec{
+					Strategy: buildapi.BuildStrategy{
+						DockerStrategy: &buildapi.DockerBuildStrategy{From: imageRef("test")},
+					},
+				},
+				oldSpec: &buildapi.CommonSpec{
+					Strategy: buildapi.BuildStrategy{
+						DockerStrategy: &buildapi.DockerBuildStrategy{From: imageRef("test")},
+					},
+				},
+			},
+			args: args{fn: func(ref *kapi.ObjectReference) error {
+				t.Errorf("should not have called mutator")
+				return nil
+			}},
+			wantSpec: &buildapi.CommonSpec{
+				Strategy: buildapi.BuildStrategy{
+					DockerStrategy: &buildapi.DockerBuildStrategy{From: imageRef("test")},
+				},
+			},
+		},
```

Comment:
this looks identical to the one above...extra cut+paste?

Reply:
Yes